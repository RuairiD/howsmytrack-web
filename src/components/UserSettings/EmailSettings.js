import React, { useState } from "react";
import { useMutation } from "react-query";

import { Col, Modal, Row, Typography } from "antd";
import { Div } from "lemon-reset";
import axios from "axios";
import apiRoot from "../../apiRoot";

const UPDATE_EMAIL_MUTATION = `mutation UpdateEmail($email: String!) {
    updateEmail(email: $email) {
        success
        error
    }
}`;

type EmailSettingsProps = {
    currentEmail: string,
};

const EmailSettings = ({ currentEmail }: EmailSettingsProps) => {
    const [email, setEmail] = useState(currentEmail);
    // Used to stop Modal being double shown if onChange is invoked twice for some reason.
    const [isModalVisible, setIsModalVisible] = useState(false);

    const sendUpdateEmailRequest = (newEmail) => (
        axios.post(`${apiRoot}/graphql/`, {
            query: UPDATE_EMAIL_MUTATION,
            variables: {
                email: newEmail,
            },
        })
    );

    const [updateEmailRequestMutate, { data, reset }] = useMutation(sendUpdateEmailRequest);

    const onEditStart = () => {
        reset();
    };

    const onSubmit = (newEmail) => {
        if (newEmail === email && !isModalVisible) {
            return;
        }
        setIsModalVisible(true);
        Modal.confirm({
            title: `Are you sure you want to update your email address to "${newEmail}"?`,
            content: "You will be logged out and required to log back in with your new email address.",
            onOk: () => {
                setEmail(newEmail);
                updateEmailRequestMutate(newEmail);
                setIsModalVisible(false);
            },
            onCancel: () => {
                setIsModalVisible(false);
            },
            okText: "Yes, change it.",
            cancelText: "No, take me back.",
        });
    };

    // Log the user out and ask them to log back in with
    // their new email address.
    if (data && data.success) {
        axios.get(`${apiRoot}/logout/`).then(() => window.location.assign("/"));
    }

    return (
        <Div>
            <Row>
                <Col>
                    <Typography.Text strong>
                        Your email address:
                    </Typography.Text>
                </Col>
            </Row>
            <Row>
                <Col>
                    <Typography.Text
                        className="email"
                        style={{
                            marginRight: "1em",
                        }}
                        editable={{
                            onStart: onEditStart,
                            onChange: onSubmit,
                        }}
                    >
                        {email}
                    </Typography.Text>
                    {data && data.success && (
                        <Typography.Text type="secondary">
                            Saved
                        </Typography.Text>
                    )}
                    {data && data.error && (
                        <Typography.Text type="danger">
                            Error: {data.error}
                        </Typography.Text>
                    )}
                </Col>
            </Row>
        </Div>
    );
};

export default EmailSettings;
