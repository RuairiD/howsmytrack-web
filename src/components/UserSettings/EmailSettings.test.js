import React from "react";
import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import axios from "axios";
import { act } from "react-dom/test-utils";
import { Modal } from "antd";
import waitForExpect from "wait-for-expect";
import EmailSettings from "./EmailSettings";

const modalConfirmSpy = jest.spyOn(Modal, "confirm");

jest.mock("axios");
jest.mock("../../apiRoot", () => "http://localhost:8000");

describe("EmailSettings", () => {
    afterEach(() => {
        axios.get.mockRestore();
        axios.post.mockRestore();
        modalConfirmSpy.mockClear();
    });

    it("renders an editable text field with the user's current email address", () => {
        const wrapper = shallow(
            <EmailSettings
                currentEmail="alemac@brightonandhovealbion.com"
            />,
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("doesn't do anything if user attempts to change email address without actually changing its value", () => {
        const wrapper = mount(
            <EmailSettings
                currentEmail="alemac@brightonandhovealbion.com"
            />,
        );

        act(() => wrapper.find("Text.email").get(0).props.editable.onStart());
        act(() => wrapper.find("Text.email").get(0).props.editable.onChange(
            "alemac@brightonandhovealbion.com",
        ));

        expect(modalConfirmSpy).not.toHaveBeenCalled();

        wrapper.unmount();
    });

    it("shows a confirmation modal when changing the user's email address", () => {
        const wrapper = mount(
            <EmailSettings
                currentEmail="alemac@brightonandhovealbion.com"
            />,
        );

        act(() => wrapper.find("Text.email").get(0).props.editable.onStart());
        act(() => wrapper.find("Text.email").get(0).props.editable.onChange(
            "ltrossard@brightonandhovealbion.com",
        ));

        expect(modalConfirmSpy).toHaveBeenCalled();
        expect(modalConfirmSpy.mock.calls.length).toBe(1);

        wrapper.unmount();
    });

    it("doesn't do anything if user dismisses the confirmation modal", async () => {
        const wrapper = mount(
            <EmailSettings
                currentEmail="alemac@brightonandhovealbion.com"
            />,
        );

        act(() => wrapper.find("Text.email").get(0).props.editable.onStart());
        act(() => wrapper.find("Text.email").get(0).props.editable.onChange(
            "ltrossard@brightonandhovealbion.com",
        ));

        await act(async () => modalConfirmSpy.mock.calls[0][0].onCancel());

        expect(axios.post).not.toHaveBeenCalled();

        wrapper.unmount();
    });

    it("submits a request to change the email address if the user confirms the modal and redirects the user to the home page to re-login", async () => {
        axios.post.mockImplementationOnce(() => Promise.resolve({
            data: {
                data: {
                    updateEmail: {
                        success: true,
                        error: false,
                    },
                },
            },
        }));
        axios.get.mockImplementationOnce(() => Promise.resolve({}));

        const assign = jest.fn();
        Object.defineProperty(
            window.location,
            "assign",
            {
                value: assign,
                writable: true,
                configurable: true,
            },
        );

        const wrapper = mount(
            <EmailSettings
                currentEmail="alemac@brightonandhovealbion.com"
            />,
        );

        act(() => wrapper.find("Text.email").get(0).props.editable.onStart());
        act(() => wrapper.find("Text.email").get(0).props.editable.onChange(
            "ltrossard@brightonandhovealbion.com",
        ));

        await act(async () => modalConfirmSpy.mock.calls[0][0].onOk());

        expect(axios.post).toHaveBeenCalled();
        expect(wrapper.text().includes("Saved")).toBeTruthy();

        await waitForExpect(() => {
            wrapper.update();
            expect(axios.get).toHaveBeenCalled();
            expect(assign).toHaveBeenCalled();
            expect(assign.mock.calls[0][0]).toBe("/");
        });

        wrapper.unmount();
    });

    it("shows an error if one is returned e.g. if the email is already in use", async () => {
        const error = "email already in use";
        axios.post.mockImplementationOnce(() => Promise.resolve({
            data: {
                data: {
                    updateEmail: {
                        success: false,
                        error,
                    },
                },
            },
        }));

        const wrapper = mount(
            <EmailSettings
                currentEmail="alemac@brightonandhovealbion.com"
            />,
        );

        act(() => wrapper.find("Text.email").get(0).props.editable.onStart());
        act(() => wrapper.find("Text.email").get(0).props.editable.onChange(
            "ltrossard@brightonandhovealbion.com",
        ));

        await act(async () => modalConfirmSpy.mock.calls[0][0].onOk());

        expect(axios.post).toHaveBeenCalled();
        expect(wrapper.text().includes(error)).toBeTruthy();

        wrapper.unmount();
    });
});
