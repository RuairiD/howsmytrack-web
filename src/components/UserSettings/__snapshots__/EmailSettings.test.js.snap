// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailSettings renders an editable text field with the user's current email address 1`] = `
<Div
  className=""
>
  <Row
    gutter={0}
  >
    <Col>
      <Text
        strong={true}
      >
        Your email address:
      </Text>
    </Col>
  </Row>
  <Row
    gutter={0}
  >
    <Col>
      <Text
        className="email"
        editable={
          Object {
            "onChange": [Function],
            "onStart": [Function],
          }
        }
        style={
          Object {
            "marginRight": "1em",
          }
        }
      >
        alemac@brightonandhovealbion.com
      </Text>
    </Col>
  </Row>
</Div>
`;
